### PR TITLE
Training Infrastructure for Semantic Segmentation using Data Extractor

### PR DESCRIPTION
**Pull request by mpiseno**
## Motivation and Context

Note: A lot of this code is from my other PRs but is still being included in here for some reason.

Set up the rest of the training infrastructure for running semantic/instance segmentation models on data taken from the image extraction API. This includes evaluation metrics, visualization tools, support for extracting images from multiple scenes, and a caching data structure to prevent so many scene switches in the case where the user is using multiple scenes.

## How Has This Been Tested

I created a jupyter notebook (not included) and overfit a Mask R CNN model on a small training set. I also trained on a larger dataset that included images taken from multiple scenes and observed the loss converging. More rigorous testing can be implemented in the future, just looking for code feedback right now.

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by mathfac** [comment]
As I understand, https://github.com/pytorch/vision/tree/master/references/detection was used as reference point. Unfortunately,  `references/detection` isn't part of `torchvision` release. How significantly code was changed compare to original?
One way to address this can be adding `torchvision` repo as submodule or clone and make branch with `references/detection` dir only. Then import the `references/detection` and add only delta to `examples`.

File: habitat_sim/utils/data/data_extractor.py [hunk 1]
```diff
@@ -99,10 +141,22 @@ def __getitem__(self, idx):
             out_name: obs[self.out_name_to_sensor_name[out_name]]
             for out_name in self.output
         }
-        sample["label"] = self.label_map[label]
+        sample["label"] = label
+        self.cache.add(cache_entry, sample)
 
         return sample
 
+    def precomute_tdv_and_refs(self, sim, scene_filepaths, pixels_per_meter):
```
Comment (mathfac) [inline habitat_sim/utils/data/data_extractor.py]: what is `tdv`? please add docstring. Changes to `data_extractor.py ` can be separate PR.

File: habitat_sim/utils/data/data_extractor.py [hunk 2]
```diff
@@ -31,15 +36,28 @@ class ImageExtractor:
 
     def __init__(
         self,
-        scene_filepath,
+        filepath,
```
Comment (mathfac) [inline habitat_sim/utils/data/data_extractor.py]: maybe `scenes_filepath`, `scene` context is important.

Comment (mpiseno) [reply habitat_sim/utils/data/data_extractor.py]: Fixing this up now

File: habitat_sim/utils/data/data_extractor.py [hunk 3]
```diff
@@ -50,19 +68,31 @@ def __init__(
 
         self.sim = sim
         self.pixels_per_meter = 0.1
```
[the remark this answers is not in the document]
Comment (mpiseno) [reply habitat_sim/utils/data/data_extractor.py]: Where should I add the instructions on how to incorporate vision/references/segmentation code? I can make another tutorial on the website that explains this.

File: habitat_sim/utils/data/data_extractor.py [hunk 4]
```diff
@@ -111,6 +165,51 @@ def close(self):
...
+            print(f"House center:{scene.aabb.center} dims:{scene.aabb.sizes}")
+
+        instance_id_to_name = {}
+        for obj in scene.objects:
+            if obj and obj.category:
+                obj_id = int(obj.id.split("_")[-1])
+                instance_id_to_name[obj_id] = obj.category.name()
+
+        return instance_id_to_name
+
```
[the remark this answers is not in the document]
Comment (mpiseno) [reply habitat_sim/utils/data/data_extractor.py]: This TopdownView might be useful for other things in the future. For example, we might want to create topdown views with different pixel values corresponding to different semantic classes that would help with extracting images of specific classes. That's why I made it a separate python class.

File: habitat_sim/utils/data/data_extractor.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mpiseno) [reply habitat_sim/utils/data/data_extractor.py]: I will update the tutorial on the website with instructions on how to run the example training pipeline and how to download the files from vision/references/detection

File: habitat_sim/utils/data/data_structures.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
+class LinkedListNode:
```
Comment (erikwijmans) [inline habitat_sim/utils/data/data_structures.py]: This can be done with an OrderedDict like here: https://github.com/tkem/cachetools/blob/master/cachetools/lru.py

File: habitat_sim/utils/filesystem.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import os
+
+
+def search_dir_tree_for_ext(mydir, ext):
```
Comment (erikwijmans) [inline habitat_sim/utils/filesystem.py]: This can be done with `glob.glob(f"*.{ext}")` -- can also be done recursively like `glob.glob(f"**/*.{ext}", recursive=True)`

File: habitat_sim/utils/data/data_structures.py (hunk 5 above)
Comment (mpiseno) [reply habitat_sim/utils/data/data_structures.py]: That was a lot simpler, didn't know that existed! Fixed now.

File: habitat_sim/utils/filesystem.py (hunk 6 above)
Comment (mpiseno) [reply habitat_sim/utils/filesystem.py]: I ended up not using this function in my next PR so I got rid of it

**Comment by ventusff** [comment]
hi, there are some missing files for running the example:
```
from examples.instance_segmentation import utils
from examples.instance_segmentation.coco_eval import CocoEvaluator
from examples.instance_segmentation.coco_utils import get_coco_api_from_dataset
```
i.e. 
```
examples/instance_segmentation/utils.py
examples/instance_segmentation/coco_eval.py
examples/instance_segmentation/coco_utils.py
```

**Comment by mpiseno** [comment]
> hi, there are some files missing for running the example:
> 
> ```
> from examples.instance_segmentation import utils
> from examples.instance_segmentation.coco_eval import CocoEvaluator
> from examples.instance_segmentation.coco_utils import get_coco_api_from_dataset
> ```
> 
> i.e.
> 
> ```
> examples/instance_segmentation/utils.py
> examples/instance_segmentation/coco_eval.py
> examples/instance_segmentation/coco_utils.py
> ```

Hi, those files are a dependency for the instance segmentation training. Please refer to this tutorial, that uses the same files: https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html. They are not part of the torchvision package, which is why you need to download them yourself. Let me know if you have any other question!

**Comment by ventusff** [comment]
hi @mpiseno 
Thx for the reply! everything works like a charm now.

But I have some additional ideas that may be helpful:
As in habitat's tradition, top-down-view only draws a navigable 2d map at a _SPECIFIT HEIGHT_ (of course with some tolerance on the heights).
But most of the houses in habitat's datasets( mp3d, Gibson, or replica) are multi-level, thus extracting poses from top-down-view at a certain point only generates points from a part of the house, not the entire house. 

To settle this, you can try sampling points from a 3d-topological-graph of the house, instead of a top-down view.
Take gibson_v1's _Albertville_ for example: 
this is the _pose extractor's topdown view_: (showing only the ground floor of the house)


This is a _3d topological graph_ I've just built aiming to generate vision-task datasets:
(it's a gif, rotating)
topo2
you can see that the pose_extractor's topdown view is only the ground floor, while there are also a basement and a second floor in the house.
The topological dataset graph is also built with discretizing the world, the same as the top-down view. Each node contains the robot pose in the world coordinates.

Currently, maybe sampling top-down-view**s** from different heights for the same house might help, but it may have limitations.

**Comment by dhruvbatra** [comment]
Hi @ventusff -- this is an interesting utility, but unrelated to this thread. Would you mind sending the 3D map generation utility as a PR similar to #718? 

That way, it will be available to everyone to use (including via h-api as in https://github.com/facebookresearch/habitat-api/pull/442).

**Comment by ventusff** [comment]
> Hi @ventusff -- this is an interesting utility, but unrelated to this thread. Would you mind sending the 3D map generation utility as a PR similar to #718?
> 
> That way, it will be available to everyone to use (including via h-api as in [facebookresearch/habitat-api#442](https://github.com/facebookresearch/habitat-api/pull/442)).

Hi @dhruvbatra 
Yeah, sure! Glad to have the opportunity to make some contributions for everyone :) This component is a part of a WIP paper's repo that is going to be made public later. I'll send a PR once it's done.
